File: smartcontract/programs/doublezero-telemetry/src/pda.rs
use crate::seeds::{SEED_DEVICE_LATENCY_SAMPLES, SEED_INTERNET_LATENCY_SAMPLES, SEED_PREFIX};
use solana_program::pubkey::Pubkey;

/// Derive PDA for DZ latency samples account.
pub fn derive_device_latency_samples_pda(
    program_id: &Pubkey,
    origin_device_pk: &Pubkey,
    target_device_pk: &Pubkey,
    link_pk: &Pubkey,
    epoch: u64,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            SEED_PREFIX,
            SEED_DEVICE_LATENCY_SAMPLES,
            origin_device_pk.as_ref(),
            target_device_pk.as_ref(),
            link_pk.as_ref(),
            &epoch.to_le_bytes(),
        ],
        program_id,
    )
}

/// Derive PDA for Internet latency samples account
pub fn derive_internet_latency_samples_pda(
    program_id: &Pubkey,
    data_provider_name: &str,
    origin_location_pk: &Pubkey,
    target_location_pk: &Pubkey,
    epoch: u64,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            SEED_PREFIX,
            SEED_INTERNET_LATENCY_SAMPLES,
            data_provider_name.as_bytes(),
            origin_location_pk.as_ref(),
            target_location_pk.as_ref(),
            &epoch.to_le_bytes(),
        ],
        program_id,
    )
}

File: smartcontract/sdk/go/telemetry/pda.go
package telemetry

import (
	"encoding/binary"

	"github.com/gagliardetto/solana-go"
)

// Derives the PDA for device latency samples account
func DeriveDeviceLatencySamplesPDA(
	programID solana.PublicKey,
	originDevicePK solana.PublicKey,
	targetDevicePK solana.PublicKey,
	linkPK solana.PublicKey,
	epoch uint64,
) (solana.PublicKey, uint8, error) {
	// Convert epoch to little-endian bytes
	epochBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(epochBytes, epoch)

	// Create seeds
	seeds := [][]byte{
		[]byte(TelemetrySeedPrefix),
		[]byte(DeviceLatencySamplesSeed),
		originDevicePK[:],
		targetDevicePK[:],
		linkPK[:],
		epochBytes,
	}

	// Find program address
	return solana.FindProgramAddress(seeds, programID)
}

// Derives the PDA for internet latency samples account
func DeriveInternetLatencySamplesPDA(
	programID solana.PublicKey,
	dataProviderName string,
	originLocationPK solana.PublicKey,
	targetLocationPK solana.PublicKey,
	epoch uint64,
) (solana.PublicKey, uint8, error) {
	// Convert epoch to little-endian bytes
	epochBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(epochBytes, epoch)

	// Create seeds
	seeds := [][]byte{
		[]byte(TelemetrySeedPrefix),
		[]byte(InternetLatencySamplesSeed),
		[]byte(dataProviderName),
		originLocationPK[:],
		targetLocationPK[:],
		epochBytes,
	}

	// Find program address
	return solana.FindProgramAddress(seeds, programID)
}
